repose-aggregator/core/repose-core/src/main/scala/org/openrepose/powerfilter/ReposeFilter.scala

```diff
@@ -19,7 +19,195 @@
+
+  override def init(filterConfig: FilterConfig): Unit = {
+    logger.info("{} -- Initializing ReposeFilter", nodeId)
+    reposeFilterLoader.setServletContext(filterConfig.getServletContext)
+    logger.trace("{} -- initialized.", nodeId)
+  }
+
+  override def doFilter(request: ServletRequest, response: ServletResponse, chain: FilterChain): Unit = {
+    reposeFilterLoader.getFilterContextList match {
+      case Some(filterContextList) =>
+        logger.trace("ReposeFilter processing request...")
+        val startTime = System.currentTimeMillis
+        if (Option(request.asInstanceOf[HttpServletRequest].getHeader(TRACE_REQUEST)).isDefined) {
+          MDC.put(TRACE_REQUEST, "true")
+        }
+
+        val requestBodyInputStream =
+          Option(containerConfigurationService
+            .getContentBodyReadLimit
+            .orElse(null)) match {
+            case Some(readLimit) => new LimitedReadInputStream(readLimit, request.getInputStream)
+            case _ => request.getInputStream
+          }
+
+        val wrappedResponse = new HttpServletResponseWrapper(response.asInstanceOf[HttpServletResponse], MUTABLE, MUTABLE)
+        val bufferedInputStream = new BufferedServletInputStream(requestBodyInputStream)
+
+        // Since getParameterMap may read the body, we must reset the InputStream so that we aren't stripping
+        // the body when form parameters are sent.
+        bufferedInputStream.mark(Integer.MAX_VALUE)
+        val paramaterMap = new HttpServletRequestWrapper(request.asInstanceOf[HttpServletRequest], bufferedInputStream).getParameterMap
+        bufferedInputStream.reset()
+
+        // Wrapping the request to reset the inputStream/Reader flag
+        val wrappedRequest = new HttpServletRequestWrapper(request.asInstanceOf[HttpServletRequest], bufferedInputStream)
+
+        // Added so HERP has the Query Params available for logging.
+        wrappedRequest.setAttribute(QUERY_PARAMS, paramaterMap)
+
+        // Add the start time to be used by the ResponseTimeHandler/HttpLogFormatter.
+        wrappedRequest.setAttribute(START_TIME_ATTRIBUTE, startTime)
+
+        val scope = startSpan(wrappedRequest, tracer, logger.underlying, Tags.SPAN_KIND_CLIENT, reposeVersion, uriRedactionService)
+
+        // Conditionally remove the tracing header so it will be overwritten
+        val tracingHeaderConfig = reposeFilterLoader.getTracingHeaderConfig
+        if (tracingHeaderConfig.exists(_.isRewriteHeader)) wrappedRequest.removeHeader(TRACE_GUID)
+
+        //Grab the traceGUID from the request if there is one, else create one
+        val traceGUID = Option(wrappedRequest.getHeader(TRACE_GUID)) match {
+          case Some(string) if StringUtils.isNotBlank(string) => string
+          case _ => UUID.randomUUID.toString
+        }
+
+        MDC.put(TracingKey.TRACING_KEY, traceGUID)
+
+        try {
+          TryWith(filterContextList)(_ => {
```

Comment:
I do like the `try-with-resources` construct and Scala's `Try` in general so I chose the second option with one tweak.
I moved the `wrappedResponse.commitToResponse()` into the `Success` handler so the `Failure` handlers wouldn't have to cause an `uncommit`.

Reply:
Good call. I guess I was more focused on the structure than that tidbit.

Reply:
Actually I added it below the `doFilterChain match` block so the `sendError`'s were committed.

Reply:
Now the `Failure(t: Throwable)` and directly after the `match` both call the new `setViaClearCommitAndClose` method.

Reply:
I thought about this some more and tried to get rid of the  `setViaClearCommitAndClose` method. I came up with something that uses Scala's native `Try` rather than the `TryWith` construct. It also holds off on clearing the `MDC` until after the last possible log message, which ensures that logging occurs as expected. I was pretty explicit (with comments) about what everything was doing, so it's not as succinct as possible, but should be relatively clear. My favorite part is that there aren't multiple exit points -- regardless of what occurs during processing, it alwauys . So here is another option for you:
```
val doFilterChain = Try {
  // Ensure the request URI is a valid URI
  // This object is only being created to ensure its validity.
  // So it is safe to suppress warning squid:S1848
  new URI(wrappedRequest.getRequestURI)

  if (tracingHeaderConfig.exists(_.isEnabled)) {
    if (StringUtils.isBlank(wrappedRequest.getHeader(TRACE_GUID))) {
      wrappedRequest.addHeader(TRACE_GUID, TracingHeaderHelper.createTracingHeader(traceGUID, wrappedRequest.getHeader(VIA)))
    }
    if (tracingHeaderConfig.exists(_.isSecondaryPlainText)) {
      TraceIdLogger.trace("Adding plain text trans id to request: {}", traceGUID)
      wrappedRequest.replaceHeader(REQUEST_ID, traceGUID)
    }
    val tracingHeader = wrappedRequest.getHeader(TRACE_GUID)
    TraceIdLogger.info("Tracing header: {}", TracingHeaderHelper.decode(tracingHeader))
    wrappedResponse.addHeader(TRACE_GUID, tracingHeader)
  }

  new ReposeFilterChain(
    filterContextList.filterContexts,
    chain,
    filterContextList.bypassUriRegex,
    optMetricRegistry,
    tracer
  ).doFilter(wrappedRequest, wrappedResponse)

  logger.trace("{} -- Successfully processed request", nodeId)
}

// Release the filterContextList
// And set the Via header
filterContextList.close()
if (wrappedResponse.isCommitted) {
  wrappedResponse.uncommit()
}
responseHeaderService.setVia(wrappedRequest, wrappedResponse)

// Handle Throwables that arose while processing
doFilterChain.recover {
  case use: URISyntaxException =>
    logger.debug(s"$nodeId -- Invalid URI requested: ${wrappedRequest.getRequestURI}", use)
    wrappedResponse.sendError(SC_BAD_REQUEST, "Error processing request")
  case e: Exception =>
    logger.error(s"$nodeId -- Issue encountered while processing filter chain.", e)
    wrappedResponse.sendError(SC_BAD_GATEWAY, "Error processing request")
  case t: Throwable =>
    logger.error(s"$nodeId -- Error encountered while processing filter chain.", t)
    wrappedResponse.sendError(SC_BAD_GATEWAY, "Error processing request")
    throw t
}

// Commit the response and record metrics for the request and response
wrappedResponse.commitToResponse()
closeSpan(wrappedResponse, scope)
optMetricRegistry.foreach { mr =>
  markResponseCodeHelper(
    mr,
    wrappedResponse.getStatus,
    System.currentTimeMillis - startTime,
    logger.underlying)
}

// Either log the final message and clear out the logging context
// Or rethrow any Throwables that arose while processing and were not handled
doFilterChain.map { _ =>
  logger.trace("ReposeFilter returning response...")
  MDC.clear()
}.get
```

Reply:
@dmnjohns code doesn't quite work, he's fixing it, but that does read pretty straight forward.

Reply:
You could also lose a few LOC if you chain the `.recover` block and always `uncommit` then `setVia` afterwards. Then you always have to `uncommit`, but you get to compress things a bit more.

Reply:
To go back to using the `try/catch/finally` block approach, you can get pretty close to the same semantics with fewer LOC:
```
try {
  // Ensure the request URI is a valid URI
  // This object is only being created to ensure its validity.
  // So it is safe to suppress warning squid:S1848
  new URI(wrappedRequest.getRequestURI)

  if (tracingHeaderConfig.exists(_.isEnabled)) {
    if (StringUtils.isBlank(wrappedRequest.getHeader(TRACE_GUID))) {
      wrappedRequest.addHeader(TRACE_GUID, TracingHeaderHelper.createTracingHeader(traceGUID, wrappedRequest.getHeader(VIA)))
    }
    if (tracingHeaderConfig.exists(_.isSecondaryPlainText)) {
      TraceIdLogger.trace("Adding plain text trans id to request: {}", traceGUID)
      wrappedRequest.replaceHeader(REQUEST_ID, traceGUID)
    }
    val tracingHeader = wrappedRequest.getHeader(TRACE_GUID)
    TraceIdLogger.info("Tracing header: {}", TracingHeaderHelper.decode(tracingHeader))
    wrappedResponse.addHeader(TRACE_GUID, tracingHeader)
  }

  new ReposeFilterChain(
    filterContextList.filterContexts,
    chain,
    filterContextList.bypassUriRegex,
    optMetricRegistry,
    tracer
  ).doFilter(wrappedRequest, wrappedResponse)

  logger.trace("{} -- Successfully processed request", nodeId)
} catch {
  case use: URISyntaxException =>
    logger.debug(s"$nodeId -- Invalid URI requested: ${wrappedRequest.getRequestURI}", use)
    wrappedResponse.sendError(SC_BAD_REQUEST, "Error processing request")
  case e: Exception =>
    logger.error(s"$nodeId -- Issue encountered while processing filter chain.", e)
    wrappedResponse.sendError(SC_BAD_GATEWAY, "Error processing request")
  case t: Throwable =>
    logger.error(s"$nodeId -- Error encountered while processing filter chain.", t)
    wrappedResponse.sendError(SC_BAD_GATEWAY, "Error processing request")
    throw t
} finally {
  filterContextList.close()
  if (wrappedResponse.isCommitted) {
    wrappedResponse.uncommit()
  }
  responseHeaderService.setVia(wrappedRequest, wrappedResponse)
  wrappedResponse.commitToResponse()
  closeSpan(wrappedResponse, scope)
  optMetricRegistry.foreach { mr =>
    markResponseCodeHelper(
      mr,
      wrappedResponse.getStatus,
      System.currentTimeMillis - startTime,
      logger.underlying)
  }
  MDC.clear()
}

logger.trace("ReposeFilter returning response...")
```

The caveat here is that the final log message will not have the MDC contents intact when it is logged.